### ISSUE TITLE
Independent `Show` package

### ISSUE THREAD

**Issue by liyishuai** [QuickChick/QuickChick#168]
Per previous discussion, we should provide `Show` typeclass as a generic package (say, `coq-show`?) for more scenarios than QuickChick.

---
Maybe it's time to reorganize the existing issues: some might fit into [QuickChick 2](/QuickChick/QuickChick/projects/2) project.

**Comment by Lysxia** [comment]
It would also be nice to make `Show` produce S-expressions (which can also be parsed as Gallina expressions) rather than strings. It decouples serialization from layout (pretty-printing) and makes parsing much easier.

**Comment by Lysxia** [comment]
I created https://github.com/Lysxia/coq-ceres

**Comment by lemonidas** [comment]
ExtLib also provides a [Show module](https://github.com/coq-ext-lib/coq-ext-lib/blob/master/theories/Programming/Show.v). Any thoughts on that?

**Comment by Lysxia** [comment]
I really think we should move away from flat strings, they make our life harder than necessary. They are annoying to write (deriving doesn't cover all cases, yet), and tightly coupled with formatting concerns (even when deriving works, we still can't do much when the output is one giant string). That's why I'm pushing for S-expressions. ext-lib makes things even more complicated by using a final encoding of strings with little to no benefit.

I'm also not a fan of big libraries in the style of ext-lib in general: if we want to use a new feature, we must catch up with any breaking changes that may have happened in unrelated parts of the library.

**Comment by liyishuai** [comment]
Deriving `Show` in ExtLib seems straightforward if I understood the current derivation correctly.
I'm in favor of S-expressions, as Coq strings have made my hands really dirty.

**Comment by lemonidas** [comment]
I agree that S-expressions do seem like a straightforward target for the derivation procedure.

> I'm also not a fan of big libraries in the style of ext-lib in general: if we want to use a new feature, we must catch up with any breaking changes that may have happened in unrelated parts of the library.

Not really relevant to this issue, but does that mean you'd be in favor of not using the ExtLib monadic notation on QuickChick 2?

**Comment by Lysxia** [comment]
I would stick with extlib for the monad class for now, and wait to split when the API can be substantially improved at the same time.

There is definitely a lot of room to improve on extlib's interface (even the laws are still commented out as of now), but I run into too many questions every time I look into it, and I don't want to have to fork it again only a few months later. To me, the question of adapting the `Monad` class from Haskell to a theorem prover like Coq (with all the concerns that adds) still seems like a pretty open research problem (overlapping with the whole debate on type class vs canonical structures).

**Comment by liyishuai** [comment]
I've made `Show` a separate library: https://github.com/liyishuai/coq-show licensed under BSD-3 and proposed it to coq-community. Are there concerns regarding community maintenance @lemonidas?

**Comment by lemonidas** [comment]
Should QuickChick be made to depend on that? As long as the Coq CI is happy, I'm happy.  The current plan seems to be to only move the "Show" definitions but not the derivation. Does that mean we keep the derivation in QuickChick and hope the coq-show API doesn't change/break stuff?

What about @Lysxia's S-expression library?

**Comment by liyishuai** [comment]
It's not necessary to depend QuickChick on Show, unless the community makes some significant improvement to Show. Currently my main purpose for a separate library is that users needn't install an entire QuickChick just for a Show class.
IIUC S-expression is for QuickChick 2?

**Comment by Lysxia** [comment]
The point of S-expressions is that a generic `show` should produce tree-shaped data rather than raw strings. So coq-ceres is very much intended to make coq-show unnecessary.

**Comment by liyishuai** [comment]
The `to_string` in Ceres doesn't seem suitable for printing counterexamples in QuickChick.
Is there an agenda for your pretty-printer https://github.com/coq-community/coq-ext-lib/issues/75#issuecomment-543215319?
